### DISCUSSION TITLE
[Feature] be able to hide the next version

### DISCUSSION THREAD

**Discussion by MTroian94** [vercel/next.js#72486] [Ideas]
### Link to the code that reproduces this issue

https://github.com/MTroian94/next-version-window

### To Reproduce

1. Build the application (pnpm run build)
2. Start the production server (pnpm run start)
3. Open the browser and visit the page
4. Open the browser console
5. window.next shows the actual version

### Current vs. Expected behavior

Following the steps from the previous section, I expected to see the next version in `window.next`

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: x64
  Version: Darwin Kernel Version 22.6.0: Wed Jul  5 22:21:56 PDT 2023; root:xnu-8796.141.3~6/RELEASE_X86_64
  Available memory (MB): 16384
  Available CPU cores: 12
Binaries:
  Node: 20.9.0
  npm: 10.1.0
  Yarn: 1.22.19
  pnpm: 9.1.2
Relevant Packages:
  next: 15.0.3-canary.9 // Latest available version is detected (15.0.3-canary.9).
  eslint-config-next: N/A
  react: 19.0.0-rc-66855b96-20241106
  react-dom: 19.0.0-rc-66855b96-20241106
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next build (local)

### Additional context

These discussions:
- https://github.com/vercel/next.js/discussions/68058
- https://github.com/vercel/next.js/discussions/39488

show how to hide the Next.js version by modifying the built code. Is it possible to hide the Next.js version using the configuration in the `next.config.js` file? Setting `poweredByHeader` to false is not enough.

This is still a potential security issue for apps using Next.js, as attackers can see which Next.js version is used and potentially find exploits for that particular version.

Thank you!

**Comment by samcx** [comment]
@MTroian94 Moving this to discussions since this is a feature request. Let me discuss this with the team and get back to you!

**Reply by samcx** [reply]
@MTroian94 This was discussed to lengths in the team and it was decided that hiding the Next.js version does not realistically increase security (e.g., React itself does not hide the version).

**Reply by MTroian94** [reply]
Ok, thank you and best regards!

**Reply by mnahkies** [reply]
@samcx I think that NextJS being an actual server rather than client side renderer should have a bearing here. This gets flagged to us every time we do a pentest. 

Whilst hiding the version information is somewhat security by obscurity, it does slightly raise the difficulty of drive-by exploits, and unless there is a technical reason it needs to be present it would seem sensible to omit it.

ref: https://owasp.org/www-project-web-security-testing-guide/stable/4-Web_Application_Security_Testing/01-Information_Gathering/08-Fingerprint_Web_Application_Framework

**Reply by mnahkies** [reply]
To add a little more colour to this, there have been a reasonable number of CVEs against next recently, such as https://vercel.com/blog/postmortem-on-next-js-middleware-bypass

Exposing version information makes it much easier to scan for potentially vulnerable sites. 

Arguably you can probably still fingerprint the client side portion of the code to guess the version, but it's not as trivial as being handed the version number.

**Comment by pnalepa-bishopfox** [comment]
I believe it should be possible to hide the version number to avoid potential issues mentioned by @mnahkies

**Comment by NOOBBase** [comment]
Everyone here should treat this as priority number one.

With both the React-to-Shell exploit and the Axios exploit out in the wild, it’s honestly shocking that this hasn’t been pushed onto a “must fix immediately” level by the Next.js team.

Thousands of developers are exposed  not because they’re careless, but because bots can identify vulnerable Next.js versions in seconds. There is no proper way to obscure the version, making every app an easy target.

Let’s be real for a second:
Right now, this is practically facilitating attackers.

And before anyone falls back on the usual “just keep everything updated” argument that’s not good enough. These exploits are often disclosed after they’ve already been abused. By then, it’s too late.

This needs to be taken seriously.
If you’re not prioritizing this, you’re leaving the door wide open.

**Reply by l4rm4nd** [reply]
From the perspective of an attacker he can just throw the exploit against the nextjs instance and hope.

There is no need to confirm the version before trying an exploit. Most exploits are just a few HTTP packets.

Hiding the version is security through obscurity. I still think it would be cool to have the option to hide it but it's not that security relevant as outlined.
